### PR TITLE
Adopt vscode.diff and vscode.open for tree commands

### PR DESCRIPTION
**Pull request by alexr00**
Start with the GitFileChangeNode
Part of #2333

### CODE REVIEW

**Comment by alexr00** [comment]
@RMacfarlane I just picked a place to start. I figured it would be easier if we just to one or two uses of `vscode.diff` or `vscode.open` at a time.

**Comment by RMacfarlane** [comment]
I'm a bit confused about these changes overall, previously the commands for the tree items eventually called `vscode.diff` or `vscode.open`, does using them directly as the commands change the UI? Or is the goal to simplify the open commands?

**Comment by alexr00** [comment]
Yes, using `vscode.diff` and `vscode.open` directly now let the tree handle the commands in the same way they are handled from th explorer and SCM view. For example alt+click will open to the side. Full context is in this issue, with the decided on solution starting aproximately with this comment: https://github.com/microsoft/vscode/issues/85636#issuecomment-724844387

**Comment by alexr00** [comment]
Even after approval, I won't merge this change until insiders is updated, since it requires changes in the core to add the `TreeItem`'s `command` to the properties resolved by `resolveTreeItem`.

File: src/commands.ts [hunk 1]
```diff
@@ -586,29 +533,8 @@ export function registerCommands(context: vscode.ExtensionContext, reposManager:
...
-	}));
-	context.subscriptions.push(vscode.commands.registerCommand('pr.openChangedFile', (value: GitFileChangeNode) => {
-		const openDiff = vscode.workspace.getConfiguration().get('git.openDiffOnClick');
-		if (openDiff) {
-			return vscode.commands.executeCommand('pr.openDiffView', value);
-		} else {
-			return vscode.commands.executeCommand('review.openFile', value);
-		}
+		const command = value instanceof GitFileChangeNode ? value.openFileCommand() : openFileCommand(value);
+		vscode.commands.executeCommand(command.title, command.arguments);
```
Comment (RMacfarlane) [inline src/commands.ts:537]: typo here, should be `command.command`, and also `...` on arguments

**Review by RMacfarlane — APPROVED** [review]

File: src/view/prChangesTreeDataProvider.ts [hunk 2]
```diff
@@ -97,6 +97,13 @@ export class PullRequestChangesTreeDataProvider extends vscode.Disposable implem
 		}
 	}
 
+	async resolveTreeItem?(item: vscode.TreeItem, element: TreeNode): Promise<vscode.TreeItem> {
```
[the remark this answers is not in the document]
Comment (RMacfarlane) [reply src/view/prChangesTreeDataProvider.ts:100]: actually, how often is this called? I'm assuming just once, and then the command would get filled in then? I'm wondering if we will need to refresh the tree or something if the user changes the `openDiffOnClick` setting.

[the remark this answers is not in the document]
Comment (alexr00) [reply src/view/prChangesTreeDataProvider.ts:100]: It's new! Good catch, yes, we should refresh the tree if that setting is changed.

File: src/view/treeNodes/fileChangeNode.ts [hunk 3]
```diff
@@ -10,11 +10,30 @@ import { GitChangeType } from '../../common/file';
...
+	}
+	return {
+		command: 'vscode.open',
+		arguments: [uri, opts],
+		title: 'Open File'
+	};
+}
+
 /**
  * File change node whose content can not be resolved locally and we direct users to GitHub.
```
[the remark this answers is not in the document]
Comment (alexr00) [reply src/view/treeNodes/fileChangeNode.ts:38]: I think just making the same vscode.diff call would work. Not sure how that command handles large files.

File: src/view/treeNodes/fileChangeNode.ts [hunk 4]
```diff
@@ -162,7 +185,16 @@ export class FileChangeNode extends TreeNode implements vscode.TreeItem {
...
+		return {
+			command: 'vscode.diff',
+			arguments: [parentURI, headURI, `${pathSegments[pathSegments.length - 1]} (Pull Request)`, opts],
+			title: 'Open Changed File in PR'
+		};
+	}
+
+	async openDiff(folderManager: FolderRepositoryManager): Promise<void> {
+		const command = await this.openDiffCommand(folderManager);
+		vscode.commands.executeCommand(command.command, command.arguments);
```
[the remark this answers is not in the document]
Comment (alexr00) [reply src/view/treeNodes/fileChangeNode.ts:197]: I think the API already spreads them, so it works as is.

File: src/view/treeNodes/fileChangeNode.ts [hunk 5]
```diff
@@ -229,11 +262,83 @@ export class GitFileChangeNode extends FileChangeNode implements vscode.TreeItem
...
+
+		return {
+			command: 'vscode.diff',
+			arguments: [previousFileUri, this.filePath, `${this.fileName} from ${(commit || '').substr(0, 8)}`, options],
+			title: 'View Changes'
+		};
+	}
+
+	async resolve(): Promise<void> {
+		console.log('resolving');
```
[the remark this answers is not in the document]
Comment (alexr00) [reply src/view/treeNodes/fileChangeNode.ts:331]: 😳

File: src/commands.ts (hunk 1 above)
Comment (alexr00) [reply src/commands.ts:537]: Thanks for catching!

**Pull request merged**